src/USER-MISC/fix_charge_regulation.cpp

```diff
@@ -0,0 +1,1335 @@
+  if (seed <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (cation_type <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (anion_type <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (reaction_distance < 0.0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (salt_charge[0] <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (salt_charge[1] >= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if ((salt_charge[1] % salt_charge[0] != 0) && (salt_charge[0] % salt_charge[1] != 0))
+    error->all(FLERR,
+               "Illegal fix charge_regulation command, multivalent cation/anion charges are allowed, "
+               "but must be divisible, e.g. (3,-1) is fine, but (3,-2) is not implemented");
+
+  if (pmcmoves[0] < 0 || pmcmoves[1] < 0 || pmcmoves[2] < 0)
+    error->all(FLERR, "Illegal fix charge_regulation command");
+  if (acid_type < 0) pmcmoves[0] = 0;
+  if (base_type < 0) pmcmoves[1] = 0;
+  // normalize
+  double psum = pmcmoves[0] + pmcmoves[1] + pmcmoves[2];
+  if (psum <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  pmcmoves[0] /= psum;
+  pmcmoves[1] /= psum;
+  pmcmoves[2] /= psum;
+
+  force_reneighbor = 1;
+  next_reneighbor = update->ntimestep + 1;
+  random_equal = new RanPark(lmp, seed);
+  random_unequal = new RanPark(lmp, seed);
+  nacid_attempts = 0;
+  nacid_successes = 0;
+  nbase_attempts = 0;
+  nbase_successes = 0;
+  nsalt_attempts = 0;
+  nsalt_successes = 0;
+}
+
+Fix_charge_regulation::~Fix_charge_regulation() {
+
+  memory->destroy(ptype_ID);
+
+  delete random_equal;
+  delete random_unequal;
+
+  if (group) {
+    int igroupall = group->find("all");
+    neighbor->exclusion_group_group_delete(exclusion_group, igroupall);
+  }
+}
+
+int Fix_charge_regulation::setmask() {
+  int mask = 0;
+  mask |= PRE_EXCHANGE;
+  return mask;
+}
+
+void Fix_charge_regulation::init() {
+
+  triclinic = domain->triclinic;
+
+  char *id_pe = (char *) "thermo_pe";
+  int ipe = modify->find_compute(id_pe);
+  c_pe = modify->compute[ipe];
+
+  if (atom->molecule_flag) {
+
+    int flag = 0;
+    for (int i = 0; i < atom->nlocal; i++)
+      if (atom->type[i] == cation_type || atom->type[i] == anion_type)
+        if (atom->molecule[i]) flag = 1;
+    int flagall = flag;
+
+    MPI_Allreduce(&flag, &flagall, 1, MPI_INT, MPI_SUM, world);
+    if (flagall && comm->me == 0)
+      error->all(FLERR, "Fix charge regulation cannot exchange individual atoms (ions) belonging to a molecule");
+  }
+
+  if (domain->dimension == 2)
+    error->all(FLERR, "Cannot use fix charge regulation in a 2d simulation");
+
+  // create a new group for interaction exclusions
+  // used for attempted atom deletions
+  // skip if already exists from previous init()
+
+  if (!exclusion_group_bit) {
+    char **group_arg = new char *[4];
```

Comment:
This segment can be simplified due to recent improvements of the core LAMMPS code. Please see fix widom.